src/hostnet/lib/dns_forward.ml

```diff
@@ -14,62 +14,137 @@ module OptionThread = struct
     | Some x -> f x
 end
 
-let string_of_dns buf =
+let parse_dns buf =
   let len = Cstruct.len buf in
   let buf = Dns.Buf.of_cstruct buf in
-  match Dns.Protocol.Server.parse (Dns.Buf.sub buf 0 len) with
-  | None ->
+  (len, Dns.Protocol.Server.parse (Dns.Buf.sub buf 0 len))
+
+let string_of_dns dns =
+  match dns with
+  | (len, None) ->
     Printf.sprintf "Unparsable DNS packet length %d" len
-  | Some request ->
+  | (_, Some request) ->
     Dns.Packet.to_string request
 
+let tidstr_of_dns dns =
+  match dns with
+  | (_, None) -> "----"
+  | (_, Some { Dns.Packet.id }) -> Printf.sprintf "%04x" id
+
 module Make(Ip: V1_LWT.IPV4) (Udp:V1_LWT.UDPV4) (Resolv_conf: Sig.RESOLV_CONF) = struct
 
+type transaction = {
+  mutable resolvers : (Ipaddr.t * int) list;
+  mutable used_resolvers : (Ipaddr.t * int) list;
+  mutable last_use: float;
+}
+
+let table = Hashtbl.create 7
+
+let start_reaper () =
+  let rec loop () =
+    let open Lwt.Infix in
+    Lwt_unix.sleep 60.
+    >>= fun () ->
+    let snapshot = Hashtbl.copy table in
+    let now = Unix.gettimeofday () in
+    Hashtbl.iter (fun k trec ->
+      if now -. trec.last_use > 60. then begin
+        Log.debug (fun f -> f "DNS[%04x] Expiring DNS trec" k);
+        Hashtbl.remove table k
+      end
+    ) snapshot;
+    loop () in
+  loop ()
+
 let input ~ip ~udp ~src ~dst ~src_port buf =
   if List.mem dst (Ip.get_ip ip) then begin
 
   let src_str = Ipaddr.V4.to_string src in
   let dst_str = Ipaddr.V4.to_string dst in
 
-  Log.debug (fun f -> f "DNS %s:%d -> %s %s" src_str src_port dst_str (string_of_dns buf));
+  let dns = parse_dns buf in
 
-  (* Re-read /etc/resolv.conf on every request. This ensures that
-     changes to DNS on sleep/resume or switching networks are reflected
-     immediately. The file is very small, and parsing it shouldn't be
-     too slow. *)
-  Resolv_conf.get ()
+  let remove_tid dns =
+    match dns with
+    | (_, None) -> ()
+    | (_, Some { Dns.Packet.id }) -> Hashtbl.remove table id in
+
+  Log.debug (fun f -> f "DNS[%s] %s:%d -> %s %s" (tidstr_of_dns dns) src_str src_port dst_str (string_of_dns dns));
+
+  match dns with
+  | (_, None) -> begin
+    Resolv_conf.get ()
+    >>= function
+    | r::_ -> Lwt.return (Some r)
+    | _ -> Lwt.return None
```

Comment:
As a minor optimisation, you can use `Lwt.return_none` and `Lwt.return_some r` here which will avoid a fresh allocation of those boxes.